### ISSUE TITLE
Forcing English language error response for /api/0.6/user/details

### ISSUE THREAD

**Issue by mmd-osm** [openstreetmap/openstreetmap-website#2227]
/api/0.6/user/details is one of the very few API calls to return localized error messages. While this is good to convene some message to a mapper, in case they have been blocked, it's more difficult for editing applications like JOSM to figure out what happened.

Here's the code in question:

https://github.com/openstreetmap/openstreetmap-website/blob/master/app/controllers/api_controller.rb#L71-L84

We were wondering if it would be feasible to always include an English language text in the HTTP header field "Error", and show a localized version the HTTP body, as is today.  Would that be a reasonable thing to do?

_Follow up discussion for https://josm.openstreetmap.de/ticket/17722_

**Comment by tomhughes** [comment]
If you want a machine readable error code you should ask for that, not try and parse the text!

**Comment by mmd-osm** [comment]
Well, having machine readable error codes isn't a concept that is widely available in the API as of today. This leads to all sorts of regular expression based error message parsing, going back to even Potlatch2: https://github.com/openstreetmap/potlatch2/blob/master/net/systemeD/halcyon/connection/XMLConnection.as#L330-L364

I certainly wouldn't object putting in some more accessible error code in the "Error" HTTP header, and have the localized text in the body for the UI to show in a popup. After all, JOSM just needs the Error header to reliably recognize a "User has been blocked" situation.

**Comment by tomhughes** [comment]
Well we already have a way for forcing XML errors and then you can include whatever metadata you want.

**Comment by simonpoole** [comment]
Do we? If we do then it is not documented and "nobody" is using it. And does cgi-map support it too?

**Comment by mmd-osm** [comment]
I'm also not aware of anyone using it, at least none of the mainstream editors. CGImap has some support for HTTP_X_ERROR_FORMAT, though.

**Comment by simonpoole** [comment]
@mmd-osm ah that, but turning it on doesn't actually produce structured error messages (aka you still need to parse free form text to determine elements etc for conflicts and the like), so is rather pointless from an editor pov.

**Comment by mmd-osm** [comment]
Right. At the time of raising exceptions, all structured content has already vanished and is no longer machine readable:

```
raise OSM::APIPreconditionFailedError, "Node #{id} is still used by ways #{ways.collect(&:id).join(',')}." unless ways.empty?
```

I think it would be good to work on this at some point in the future, but that will be a lot of work.

**Comment by tomhughes** [comment]
Well sure, but it can be incremental, so we can start with the immediate case at issue here...

**Comment by simonpoole** [comment]
> I think it would be good to have more accessible error message at some point in the future, but that will be a lot of work.

I discussed this with @zerebubuth a while back and it is on his, one of many, lists for things for API 0.7, IMHO it is one of those things that would have a clear benefit and could be added in a non-"0.7" version as an intermediate step (together with a promise of semantic versioning :_)).

**Comment by mmd-osm** [comment]
While we're at it, this stuff would have to be rendered via some builder, rather than resorting to some piecemeal XML building approach. This way, it will also be easier to provide JSON response error message (which I'm sure will be the next thing to come up).

```
      result = OSM::API.new.get_xml_doc
      result.root.name = "osmError"
      result.root << (XML::Node.new("status") << "#{Rack::Utils.status_code(status)} #{Rack::Utils::HTTP_STATUS_CODES[status]}")
      result.root << (XML::Node.new("message") << message)

render :xml => result.to_s
```

**Comment by mmd-osm** [comment]
After looking into the XML topic some more, I believe it's too expensive for this use case and given the effort involved, provides too little value. 

In the meantime, I favor a solution, which gets rid of the I18N for blocked messages altogether and only returns the English text.

Reasons:
- API error messages are in English. Pretty much all of them.
- Currently, none of the translated blocked user messages ever reach a user UI, neither on osm.org, nor in any editor I checked (!)
- If editors wish to show a localized version of a blocked user message, they can do so at their own discretion.

The second point probably deserves a few words, which is I added a short analysis for setup_user_auth below:

- changesets_controller: only applies to anonymous changesets, concept no longer exists -> irrelevant
- old_controller: history+version for redacted versions: GET /api/0.6/object/* call without Authentication simply skips redacted versions. Only a call with successful authentication would return an error message for a blocked user.
- notes_controller: blocked user cannot add notes. Trying to post new notes via /note/new will trigger a POST to  /api/0.6/notes.json, which returns with 403 and an error message. However, the UI doesn't handle the error message at all, and simply gets stuck, i.e. the user never sees the error message.
- permission_controller: only relevant for editors anyway, which are expected to handle error messages.

**Comment by tomhughes** [comment]
Yes it's a long standing bug that most API messages are not properly internationalised ;-)

**Comment by mmd-osm** [comment]
Most iD users would never find out, even if they were ;)

iD tries very hard to shield all those technical messages away from a user, as they're too technical in most cases, and would be confusing to an average user.

**Comment by simonpoole** [comment]
> 
> 
> Yes it's a long standing bug that most API messages are not properly internationalised ;-)

I'm not exactly enthusiastic about adding 200 or so translated versions of already brittle regexps so I would consider that a feature.

**Comment by tomhughes** [comment]
Presumably you missed the bit about how we should have proper machine readable versions of the errors rather than people parsing the text with regexps!

**Comment by mmd-osm** [comment]
fyi: @gravitystorm mentioned this topic ("I think it would be good to have more accessible error message at some point in the future") again in his blog: https://blog.gravitystorm.co.uk/2019/10/17/smoother-api-upgrades-for-openstreetmap/
